REGRESSION(266247@main): PDF "Save" button does nothing, "Print" function also broken

mcatanzaro opened the pull request:
#### 58df23fe5ac509adc2117503127e97d53b02445b
<pre>
REGRESSION(266247@main): PDF &quot;Save&quot; button does nothing, &quot;Print&quot; function also broken
<a href="https://bugs.webkit.org/show_bug.cgi?id=260771">https://bugs.webkit.org/show_bug.cgi?id=260771</a>

Reviewed by Chris Dumez.

The PDF.js viewer protocol and resource protocols both should be treated
the same as HTTP protocols and have access to blob URLs. The protocol is
an implementation detail that shouldn&apos;t affect how the web content
behaves and certainly should not restrict access to web platform
features.

* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::isSafelistedBlobProtocol):

Canonical link: <a href="https://commits.webkit.org/270274@main">https://commits.webkit.org/270274@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b1d53f2bbd478d15d099e60810d96348ee1eec45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20931 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22404 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19135 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20737 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24189 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21106 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20531 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20724 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20588 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17827 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23257 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17756 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18632 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24920 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18833 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18808 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22860 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19405 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16483 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18617 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5995 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22958 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19224 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/page/SecurityOrigin.cpp:
```diff
@@ -169,18 +169,6 @@ Ref<SecurityOrigin> SecurityOrigin::create(const URL& url)
-    return url.protocolIsInHTTPFamily()
-        || url.protocolIsFile()
-        || LegacySchemeRegistry::schemeIsHandledBySchemeHandler(url.protocol());
-}
```
I don't think we want to remove this and regress on WPT. Additionally, not offering `blob:` URLs with the same authority is also a security win.

The original proposal of conditionally allowing the PDFJS scheme here still seems reasonable.

mcatanzaro replied:
Agreed. This is breaking WPT tests so we can't do it.

> The original proposal of conditionally allowing the PDFJS scheme here still seems reasonable.

I will allow resource:// too. I'm a little surprised we haven't noticed anything breaking due to that yet, but I guess we only use resource:// for subresources nowadays. (We used to use it for PDF.js itself before it moved from Epiphany into WebKit.)

annevk replied:
Is there a reason we'd want to create `blob:` URLs inside `resource://` documents? `blob:` URLs have been used as an attack vector in other browsers, it's better to keep them locked down.

mcatanzaro replied:
Same reason you want blob URLs to work when using `webkit-pdfjs-viewer://`. Think of `resource://` and `webkit-pdfjs-viewer://` and all other custom protocols as a replacement for HTTP. You still want the web content to function normally regardless of the protocol used to load it.

`resource://` means the content is compiled into the ELF binary rather than sent over the internet, but that's no reason to prevent blob URLs from working. Concrete example of why blob should work: before `webkit-pdfjs-viewer://` existed, Epiphany used `resource://` to load PDF.js and could hypothetically eventually want to do that again. (WebKit on Linux actually treats `webkit-pdfjs-viewer://` basically the same as `resource://`: the PDF.js content is compiled into the ELF binary using GResource and it just gets a special URL scheme.)

annevk replied:
The problem is that it allows attackers to mint these URLs as well (if there's some other bug) and they end up having the same authority. That's why you generally want to restrict it. So unless we have an actual use case I'd rather not.

mcatanzaro replied:
The use case is you can package web content into the application binary instead of having to install it in separate files on disk. So instead of WebKit using read() syscalls to read, say, 50 different files when loading an application-defined web page with 49 subresources, all of that content is already in the process address space and gets accessed using pointers instead of syscalls. GResource is an implementation detail and shouldn't affect the functionality of blob URLs.

annevk reviewed (commented):
Non-reviewer looks good.

cc @achristensen07

mcatanzaro committed:
REGRESSION(266247@main): PDF "Save" button does nothing, "Print" function also broken
https://bugs.webkit.org/show_bug.cgi?id=260771

Reviewed by Chris Dumez.

The PDF.js viewer protocol and resource protocols both should be treated
the same as HTTP protocols and have access to blob URLs. The protocol is
an implementation detail that shouldn't affect how the web content
behaves and certainly should not restrict access to web platform
features.

* Source/WebCore/page/SecurityOrigin.cpp:
(WebCore::isSafelistedBlobProtocol):

Canonical link: https://commits.webkit.org/270274@main
